File: problems/binary_search_trees/same_bsts/cpp/same_bsts.cpp
#include "same_bsts.h"
#include <utility>

bool sameBsts(const std::vector<int> &arrayOne,
              const std::vector<int> &arrayTwo) {

  if (arrayOne.empty() && arrayTwo.empty())
    return true;

  if ((arrayOne.size() != arrayTwo.size()) || (arrayOne[0] != arrayTwo[0]))
    return false;

  auto getChildren = [](const std::vector<int> &array)
      -> std::pair<std::vector<int>, std::vector<int>> {
    std::vector<int> smaller, greaterOrEqual;

    for (size_t i = 1; i < array.size(); i++) {
      if (array[i] < array[0])
        smaller.push_back(array[i]);
      else
        greaterOrEqual.push_back(array[i]);
    }

    return {smaller, greaterOrEqual};
  };

  auto [leftOne, rightOne] = getChildren(arrayOne);
  auto [leftTwo, rightTwo] = getChildren(arrayTwo);

  return sameBsts(leftOne, leftTwo) && sameBsts(rightOne, rightTwo);
}

File: problems/binary_search_trees/same_bsts/js/same_bsts.js
export const sameBsts = (() => {
  const getChildren = (array) => {
    const smaller = [];
    const greater_or_equal = [];

    for (let i = 1; i < array.length; i++) {
      if (array[i] < array[0]) {
        smaller.push(array[i]);
      } else {
        greater_or_equal.push(array[i]);
      }
    }
    return { smaller, greater_or_equal };
  };

  return (arrayOne, arrayTwo) => {
    if (!arrayOne.length && !arrayTwo.length) return true;

    if (arrayOne.length !== arrayTwo.length || arrayOne[0] !== arrayTwo[0])
      return false;

    const { smaller: leftOne, greater_or_equal: rightOne } =
      getChildren(arrayOne);
    const { smaller: leftTwo, greater_or_equal: rightTwo } =
      getChildren(arrayTwo);

    return sameBsts(leftOne, leftTwo) && sameBsts(rightOne, rightTwo);
  };
})();
